### DISCUSSION TITLE
How to draw multiple pointcould successfully one after the other in the user interface

### DISCUSSION THREAD

**Discussion by bilel-bj** [isl-org/Open3D#3725] [Q&A] [answered]
I want to build an application that:

1. It has a file selection menu to select two point clouds.
2. A button:"ICP" that starts the ICP registration algorithm between the two algorithms.
3. After pushing the ICP button:
4. it will show the two pointcloud registered and a text zone that displays the transformation matrix.

Any idea on how to do that

**Comment by ssheorey** [comment]
Hi @bilel-bj check out this C++ registration example:

https://github.com/intel-isl/Open3D/blob/master/examples/cpp/TICPOdometry.cpp

Python GUi examples are in the `examples/python/gui` folder

**Comment by vergegenkunft** [comment]
Hi @bilel-bj , did you solve this problem? I want to build a similar User Interface like yours with the funtion to show two pointclouds before and after registration, still got no idea. And i can't open the link from @ssheorey, can you update the Link, many thanks!

**Comment by ssheorey** [comment, marked as answer]
A simple way to do that is with actions in the new `draw()` function:
```py
def actions():
    SOURCE_NAME = "Source"
    RESULT_NAME = "Result (Poisson reconstruction)"
    TRUTH_NAME = "Ground truth"

    bunny = o3d.data.BunnyMesh()
    bunny_mesh = o3d.io.read_triangle_mesh(bunny.path)
    bunny_mesh.compute_vertex_normals()

    bunny_mesh.paint_uniform_color((1, 0.75, 0))
    bunny_mesh.compute_vertex_normals()
    cloud = o3d.geometry.PointCloud()
    cloud.points = bunny_mesh.vertices
    cloud.normals = bunny_mesh.vertex_normals

    def make_mesh(o3dvis):
        # TODO: call o3dvis.get_geometry instead of using bunny_mesh
        mesh, _ = o3d.geometry.TriangleMesh.create_from_point_cloud_poisson(
            cloud)
        mesh.paint_uniform_color((1, 1, 1))
        mesh.compute_vertex_normals()
        o3dvis.add_geometry({"name": RESULT_NAME, "geometry": mesh})
        o3dvis.show_geometry(SOURCE_NAME, False)

    def toggle_result(o3dvis):
        truth_vis = o3dvis.get_geometry(TRUTH_NAME).is_visible
        o3dvis.show_geometry(TRUTH_NAME, not truth_vis)
        o3dvis.show_geometry(RESULT_NAME, truth_vis)

    vis.draw([{
        "name": SOURCE_NAME,
        "geometry": cloud
    }, {
        "name": TRUTH_NAME,
        "geometry": bunny_mesh,
        "is_visible": False
    }],
             actions=[("Create Mesh", make_mesh),
                      ("Toggle truth/result", toggle_result)])
```

See the full example here: https://github.com/isl-org/Open3D/blob/master/examples/python/visualization/draw.py
